File: packages/cli/src/ckpt/commands/add.py
"""ckpt add — stage files and optionally attach reasoning."""

from __future__ import annotations

import click

from ckpt.core import git
from ckpt.storage.local import LocalStore


@click.command()
@click.argument("files", nargs=-1)
@click.option("-A", "--all", "add_all", is_flag=True, help="Stage all changes.")
@click.option("--reasoning", default=None, help="Why this change was made.")
def add(files: tuple[str, ...], add_all: bool, reasoning: str | None):
    """Stage files and optionally record reasoning."""
    # Build git add args
    git_args = ["add"]
    if add_all:
        git_args.append("--all")
    git_args.extend(files)

    if not add_all and not files:
        click.echo("Nothing specified, nothing added.")
        click.echo("Maybe you wanted to say 'ckpt add .'?")
        raise SystemExit(1)

    git.run(*git_args)

    staged = git.get_staged_files()
    if not staged:
        click.echo("No changes staged.")
        return

    click.echo(f"Staged {len(staged)} file(s)")

    if reasoning:
        store = LocalStore()
        store.add_staging(files=staged, reasoning=reasoning)
        store.close()
        click.echo(f"Reasoning saved: {reasoning}")

File: packages/cli/src/commands/add.ts
import { Command } from 'commander';
import { run, getStagedFiles } from '../core/git.js';
import { requireInit } from '../core/config.js';
import { LocalStore } from '../storage/local.js';

export const add = new Command('add')
  .description('Stage files and optionally attach reasoning')
  .argument('[files...]', 'files to stage')
  .option('-A, --all', 'stage all changes')
  .option('--reasoning <text>', 'reasoning for these changes')
  .allowUnknownOption(true)
  .action((files: string[], opts: { all?: boolean; reasoning?: string }) => {
    requireInit();

    const gitArgs = ['add'];
    if (opts.all) gitArgs.push('--all');
    if (files.length > 0) gitArgs.push(...files);
    if (!opts.all && files.length === 0) {
      console.error('Nothing to add. Use ckpt add <files> or ckpt add --all');
      process.exit(1);
    }

    const result = run(gitArgs);
    if (result.exitCode !== 0) process.exit(result.exitCode);

    if (opts.reasoning) {
      const staged = getStagedFiles();
      const store = new LocalStore();
      store.addStaging(staged, opts.reasoning);
      store.close();
      console.log(`Staged ${staged.length} file(s) with reasoning.`);
    }
  });
